var/spack/repos/builtin/packages/python/package.py

```diff
@@ -806,6 +734,61 @@ def setup_dependent_build_environment(self, env, dependent_spec):
         pythonpath = ':'.join(python_paths)
         env.set('PYTHONPATH', pythonpath)
 
+        if self.spec in dependent_spec.dependencies(deptype='link'):
+            # We want to make sure that the extensions are compiled and linked
+            # with Spack wrapper. Paths to the executables that are used for
+            # these operations are normally taken from sysconfigdata file,
+            # which we modify after the installation (see method filter
+            # compilers). The modified file contains paths to the real
+            # compilers, not the wrappers. Therefore, we back up the original
+            # sysconfigdata file before introducing any changes to it (see
+            # method backup_sysconfigdata) and prepend its path to PYTHONPATH
+            # of the dependent package.
+
+            # In several cases (see below), we need to set LDSHARED, which is
+            # used for linking, and CC, which is used for compilation, will be
+            # set by Spack automatically. Note that there are two more
+            # variables in the sysconfigdata file that are similar to LDSHARED:
+            # BLDSHARED and LDCXXSHARED. However, it looks like distutils does
+            # not check the respective environment variables. Therefore, we
+            # ignore them for now.
+            set_ldshared = True
+
+            sysconfig_dir = self.get_spack_sysconfig_dirname()
+            if os.path.isdir(sysconfig_dir):
+                env.prepend_path('PYTHONPATH', sysconfig_dir)
+                if self.spec.satisfies('@2.7.5:2.7.999,3.6.0:'):
+                    # In this case, both sysconfig and distutils.sysconfig read
+                    # data from sysconfigdata file and we do not have to do
+                    # anything else.
+                    set_ldshared = False
+
+            if set_ldshared:
+                # We need to set LDSHARED because we are dealing with:
+                #   A) installation that relies on the old logic with a JSON
+                #      file;
+                #   B) old version of python, in which distutils.sysconfig does
+                #      not use sysconfigdata file;
+                #   C) externally installed python.
+                # However, to avoid significant inconsistency, we set LDSHARED
+                # only when the respective value returned by
+                # distutils.sysconfig starts with the path to the real compiler
+                # that will be used for the compilation of the dependent
+                # package. This should sufficiently cover cases A and B, as
+                # well as case C in some rare cases.
```

Comment:
What do you mean by "in some rare cases"? I assume this would be normal when the external python was built with a different compiler than what is used for the dependent.

Reply:
To my understanding, it is safe to modify and set `LDSHARED` only when the output of the command
```console
$ python -c 'from distutils.sysconfig import get_config_var; print(get_config_var("LDSHARED"))'`
```
starts with the path to the real compiler that is used for the extension. Here is a couple of examples of the output of the command:

```console
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-z,relro -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-V2e5UW/python2.7-2.7.13=. -fstack-protector-strong -Wformat -Werror=format-security
```
The beginning of the string is not `/usr/bin/gcc`, which I might want to use in this case. Ok, we could compare the output of
```console
$ python -c "import os; print(os.path.realpath('$(which x86_64-linux-gnu-gcc)'))"
```
and
```console
$ python -c "import os; print(os.path.realpath('/usr/bin/gcc'))"
```
and see that the result is the same and still make a substitution in `LSDHARED`. But what about this output on another machine:
```console
gcc -pthread -shared
```
Can we use the same logic? Can we be sure that `gcc` returned by the `which` command is the compiler that was used for the building of Python? It might also be the case that the compiler in use is not `/usr/bin/gcc` and therefore the real paths won't match anyway. Can we be sure that the compiler in use will accept the arguments listed in `LDSHARED`? What if we are trying to build an extension using Intel compiler? There are many edge cases, especially when we are dealing with an external Python. I would rather not introduce complex logic to cover them but keep the simple one that is implemented here. However, this simple implementation will rarely work with external installations of Python because the default value of `LDSHARED` of a system installation rarely starts with the full path to the compiler in use. My recommendation for the users would be: it's fine to use an external installation of Python if it's always just a build dependency in your software tree but you should really consider building Python with Spack if you want to install compiled extensions for it.